### PR TITLE
Drop `Experimental::RawMemoryAllocationFailure` and don't catch exceptions to rethrow them in shared alloc

### PR DESCRIPTION
**Pull request by dalg24**
Motivation to get rid of `Experimental::RawMemoryAllocationFailure`:
* backends do not really distinguish failure modes and it is not clear it is useful
* the alignment is not controlled by the user, its value is static, so there is little point in attaching it to the exception
* we were actually catching them to rethrow a regular runtime error
* in the end, the most important thing is to raise an error with a good quality error message

The changes in this PR simplify the bad alloc error reporting.
New error messages are along these lines:
```
Kokkos ERROR: Host memory space failed to allocate 1.678e+07 TiB (label="hello_world_failure").
```

I also got rid of the "catch the exception, append to the error message, and rethrow" pattern in shared alloc because we have all the necessary information when we first throw.

### CODE REVIEW

File: core/unit_test/TestViewAPI.hpp [hunk 1]
```diff
@@ -1571,51 +1571,6 @@ class TestViewAPI {
...
   }
-
-  static void run_test_error() {
-// FIXME_MSVC_WITH_CUDA
-// This test doesn't behave as expected on Windows with CUDA
-#if defined(_WIN32) && defined(KOKKOS_ENABLE_CUDA)
-    if (std::is_same<typename dView1::memory_space,
-                     Kokkos::CudaUVMSpace>::value)
-      return;
-#endif
```
Comment (dalg24) [inline core/unit_test/TestViewAPI.hpp:1582]: Noting that this was never true in the current develop but I explicitly instantiated the new test on shared memory spaces so we will see if that is still an issue.

File: core/src/Cuda/Kokkos_CudaSpace.cpp [hunk 2]
```diff
@@ -221,7 +198,7 @@ void *impl_allocate_common(const int device_id,
     // we should do here since we're turning it into an
     // exception here
     cudaGetLastError();
-    throw_cuda_allocation_failure(arg_alloc_size, error_code, "cudaMalloc()");
+    Kokkos::Impl::throw_bad_alloc(arg_handle.name, arg_alloc_size, arg_label);
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_CudaSpace.cpp:201]: Could also do
```suggestion
    Kokkos::Impl::throw_bad_alloc(Kokkos::CudaSpace::name(), arg_alloc_size, arg_label);
```

File: core/src/impl/Kokkos_Error.hpp [hunk 3]
```diff
@@ -17,99 +17,20 @@
 #ifndef KOKKOS_IMPL_ERROR_HPP
 #define KOKKOS_IMPL_ERROR_HPP
 
-#include <new>  // align_val_t
 #include <string>
-#include <iosfwd>
 #include <Kokkos_Macros.hpp>
```
Comment (Rombur) [inline core/src/impl/Kokkos_Error.hpp:21]: You only need to include `<string>, the other includes are not necessary in the header file.

Comment (dalg24) [reply core/src/impl/Kokkos_Error.hpp:21]: You're correct, but a lot of headers relied on these transitive includes and I did not really want to fix it here.
Is that ok to leave it as future cleanup work?

Comment (Rombur) [reply core/src/impl/Kokkos_Error.hpp:21]: Yes, we can do it another time.

**Review by masterleinad — APPROVED** [review]
Looks good to me.

**Comment by dalg24** [comment]
Ignoring OpenMPTarget failure
```
clang++: error: no library 'libomptarget-nvptx-sm_70.bc' found in the default clang lib directory or in LIBRARY_PATH; use '--libomptarget-nvptx-bc-path' to specify nvptx bitcode library
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/build.make:104: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Command_Line_Parsing.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
clang++: error: no library 'libomptarget-nvptx-sm_70.bc' found in the default clang lib directory or in LIBRARY_PATH; use '--libomptarget-nvptx-bc-path' to specify nvptx bitcode library
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/build.make:76: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_Abort.cpp.o] Error 1
[  1%] Building CXX object _deps/googlebenchmark-build/src/CMakeFiles/benchmark.dir/benchmark_api_internal.cc.o
clang++: error: no library 'libomptarget-nvptx-sm_70.bc' found in the default clang lib directory or in LIBRARY_PATH; use '--libomptarget-nvptx-bc-path' to specify nvptx bitcode library
[  1%] Built target AlwaysCheckGit
make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/build.make:90: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_CPUDiscovery.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1561: core/src/CMakeFiles/kokkoscore.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

**Pull request merged**